pkgs/top-level/all-packages.nix

```diff
@@ -1329,7 +1329,9 @@ with pkgs;
 
   replaceDependency = callPackage ../build-support/replace-dependency.nix { };
 
-  replaceVars = callPackage ../build-support/replace-vars { };
+  replaceVars = callPackage ../build-support/replace-vars { } // {
+    withoutCheck = callPackage ../build-support/replace-vars { doCheck = false; };
+  };
```

Comment:
I don’t understand how this works. `(x: x) // { }` is an error. Is this some `__functor` magic? I think I would prefer something like `replaceVarsWith { doCheck = false; } ./file { … }`, where `replaceVars = replaceVarsWith { }`.

Reply:
> Is this some `__functor` magic?

Yes, kind of. An attrset with `__functor` is _also_ just an attrset.

```
nix-repl> replaceVars = callPackage ./pkgs/build-support/replace-vars { }

nix-repl> replaceVars
{
  __functionArgs = { ... };
  __functor = «lambda __functor @ /home/philip/Code/github.com/philiptaron/nixpkgs/lib/trivial.nix:957:19»;
  override = { ... };
}

nix-repl> replaceVars // { a = 42; }
{
  __functionArgs = { ... };
  __functor = «lambda __functor @ /home/philip/Code/github.com/philiptaron/nixpkgs/lib/trivial.nix:957:19»;
  a = 42;
  override = { ... };
}
```

It's used a couple places in `all-packages.nix`:

1. https://github.com/NixOS/nixpkgs/blob/6cc3e274c947096556c57d2b05dad5acda395b74/pkgs/top-level/all-packages.nix#L4036-L4038
2. https://github.com/NixOS/nixpkgs/blob/6cc3e274c947096556c57d2b05dad5acda395b74/pkgs/top-level/all-packages.nix#L15825 (where I learned this trick)
3. https://github.com/NixOS/nixpkgs/blob/6cc3e274c947096556c57d2b05dad5acda395b74/pkgs/top-level/all-packages.nix#L33290-L33295
4. https://github.com/NixOS/nixpkgs/blob/6cc3e274c947096556c57d2b05dad5acda395b74/pkgs/top-level/all-packages.nix#L33961-L33964

Reply:
I'll note that `callPackage` is used in cases like this where it doesn't actually resolve to a package, but rather to a factory function that will then produce packages/derivations. If there were something like a `callPackageFactory` function that included the ability to passthru additional contents, this would look quite a bit more sane.

The core piece of code that's creating this functor attrset is 

https://github.com/NixOS/nixpkgs/blob/4e2566ea2bf25af8b94d733bd610361226b09053/lib/customisation.nix#L264

As it is, I'm pretty sure that this specific feature doesn't yet justify `replaceVarsWith`, but if you feel quite strongly, I could introduce it.

Reply:
Right, I figured `callPackage` might be doing something like this. I just think `__functor` is weird and people don’t expect to find more functions under a function.

The nice thing about `replaceVarsWith` is that it would compose with any future extensions, like if we want to support `postPatch` or something. If you’re totally sure that there’ll never be another option ever again then maybe it could just be `replaceVarsWithoutCheck`?

Reply:
> people don’t expect to find more functions under a function

One of the reasons I tilted this (odd) way was that I was able to document this variant "under" `replaceVars` so that it appeared in context on https://noogle.dev/f/pkgs/replaceVars. That documentation outweighs people finding functions under a function... I think. 😁

Reply:
I agree with @emilazy that `__functor` should be avoided when possible, it can be very confusing. I like the idea of `replaceVarsWith` more, especially because the same pattern is being used elsewhere too. Check out the [docs for `lib.fileset.gitTracked`](https://nixos.org/manual/nixpkgs/stable/#function-library-lib.fileset.gitTracked) as an example how such a pair of functions can be crosslinked together in the docs, but nesting one under the other would also be fine imo.

Reply:
All right, I'll do `replaceVarsWith`.